4.x data migration (#866)
medic/cht-docs: content/en/apps/guides/hosting/4.x/data-migration.md

---
title: "Data migration from CHT 3.x to CHT 4.x"
linkTitle: "Data migration to 4.x"
weight: 10
description: >
  Guide to migrate existent data from CHT 3.x to CHT 4.x
aliases:
relatedContent: >
---

The hosting architecture differs entirely between CHT-Core 3.x and CHT-Core 4.x. Migrating data from an existing instance running CHT 3.x requires a few manual steps. 
This guide will present the required steps while using a migration helping tool, called `couchdb-migration`. This tool interfaces with CouchDb, to update shard maps and database metadata. 
Using this tool is not required, and the same result can be achieved by calling CouchDb endpoints directly. [Consult CouchDB documentation for details about moving shards](https://docs.couchdb.org/en/3.2.2-docs/cluster/sharding.html#moving-a-shard). 

### 1. Install CHT data migration tool

Open your terminal and run these commands. They will create a new directory, download a docker compose file and download the required docker image. 
```shell
mkdir -p ~/couchdb-migration/ 
cd ~/couchdb-migration/ 
curl -s -o ./docker-compose.yml https://github.com/medic/couchdb-migration/blob/main/docker-compose.yml

> docker-compose up

Author: No, this one doesn't require any env vars, as it doesn't run anything,

```

For the following steps, the tool needs access to your CouchDb installation. To allow this access, you will need to provide a URL to your CouchDB installation that includes authentication. 
Additionally, if your CouchDb runs in docker, the tool needs to be added to the same docker network in order to access protected endpoints:

```shell
export COUCH_URL=http://admin:pass@127.0.0.1:5984
```
or 
```shell
export CHT_NETWORK=<docker-network-name>
export COUCH_URL=http://admin:pass@docker-service-name:5984

> ```

Author: I think I changed all things to <place holder>. Please flag if I missed one.

For simplicity, you could store these required values in an `.env` file: 
```shell
cat > ${HOME}/couchdb-migration/.env << EOF
CHT_NETWORK=cht-net
COUCH_URL=http://admin:pass@127.0.0.1:5984
EOF
```

### 2. Prepare CHT-Core 3.x installation for upgrading
To minimize downtime when upgrading, it's advised to prepare the 3.x installation for the 4.x upgrade, and pre-index all views that are required by 4.x.

The migration tool provides a command which will download all 4.x views to your 3.x CouchDb, and initiate view indexing.

```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration pre-index-views <desired CHT version>
```

Once view indexing is finished, proceed with the next step.

{{% alert title="Note" %}} If this step is omitted, 4.x API will fail to respond to requests until all views are indexed. Depending on the size of the database, this could take many hours, or even days. {{% /alert %}} 

### 3. Save existent CouchDb configuration

Some CouchDb configuration values must be ported from existent CouchDb to the 4.x installation. Store them in a safe location before shutting down 3.x CouchDb.

##### a. CouchDB secret 

> Used in encrypting all CouchDb passwords and session tokens.

Author: The tool has a command `get-env` which will get these values, and it's explained right below. 
I switched the phrases around so we start with the description of the command and then the description of the variables. 
I don't want to provide instructions on getting them manually.

##### b. CouchDb server uuid

> Used in generating replication checkpointer documents, which track where replication progress between every client and the server, and ensure that clients don't re-download or re-upload documents.

Author: > Turning off nginx and haproxy can break the API which we may need up

API can be down for these operations, we interface directly with CouchDb. Definitely the server has to be stopped. Is that not obvious?

> Used in generating replication checkpointer documents, which track where replication progress between every client and the server, and ensure that clients don't re-download or re-upload documents.

Reviewer: I see this was added - thank you!

Use the migration tool to obtain these values.
```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration get-env
```

### 4. Locate and backup CouchDb Data folder
a) If running in MedicOS, [CouchDb data folder]({{< relref "apps/guides/hosting/3.x/self-hosting#backup" >}}) can be found at `/srv/storage/medic-core/couchdb/data`.

b) If running a custom installation of CouchDb, data would be typically stored at `/opt/couchdb/data`.

c) TODO when using orchestration / AWS 

### 5. Launch 4.x CouchDb installation

Depending on your project scalability needs and technical possibilities, you must decide whether you will deploy CouchDb in a single node or in a cluster with multiple nodes.

> Please consult this guide about clustering and horizontal scalability to make an informed decision. <insert link>

Author: I think the information about advantages/disadvantages of clustering should be somewhere else, outside of this guide. Hence the "insert link".

> Please consult this guide about clustering and horizontal scalability to make an informed decision. <insert link>

Reviewer: Fair enough - we're waiting for this [ticket then](https://github.com/medic/cht-docs/issues/831)? Works for me!

> Please consult this guide about clustering and horizontal scalability to make an informed decision. <insert link>

Reviewer: Noting we should come back and add a link to the "this guide" text - fine to leave as is for now - resolving.

{{% alert title="Note" %}} You can start with single node and then change to a cluster. This involves running the migration tool again to distribute shards from the existent node to the new nodes. {{% /alert %}}

Depending on your choice, follow the instructions that match your deployment below: 

#### Single node

a) Download 4.x single-node CouchDb docker-compose file:
```shell
mkdir -p ~/couchdb-single/ 
cd ~/couchdb-single/ 

> curl -s -o ./docker-compose.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.1.0/docker-compose/cht-couchdb.yml

Author: I would trust that people will know which version of CHT CouchDb they want to install. Serving them with latest image might not be desirable.

> curl -s -o ./docker-compose.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.1.0/docker-compose/cht-couchdb.yml

Reviewer: I can see not going latest - makes sense! 

I think we should make this be `<desired CHT version>` then so it matches the version specified in step 2 for `pre-index-views`.  Forcing them to choose so they get it right seems better instead of deciding for them and they might get it wrong.

> curl -s -o ./docker-compose.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.1.0/docker-compose/cht-couchdb.yml

Reviewer: this still seems like a worthwhile change!

> curl -s -o ./docker-compose.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.1.0/docker-compose/cht-couchdb.yml

Author: Oh, so sorry I missed this!

```
a) Make a copy of the 3.x CouchDb data folder from **step 4**.

b) Set the correct environment variables:
```shell
cat > ${HOME}/couchdb-single/.env << EOF
COUCHDB_USER=<admin>
COUCHDB_PASSWORD=<password>
COUCHDB_SECRET=<COUCHDB_SECRET from step 3>
COUCHDB_UUID=<COUCHDB_UUID from step 3>
COUCHDB_DATA=<absolute path to folder created in step 5.a>
EOF
```
c) Start 4.x CouchDb and wait until it is up.
```shell
cd ~/couchdb-single/ 
docker-compose up -d
cd ~/couchdb-migration/ 
docker-compose run couch-migration check-couchdb-up

> ```

Reviewer: the "wait until it is up" part - since we're not passing `-d` to be detached, will this naturally stop on it's own or will they need to start the next step in another terminal?  We should explicitly tell them what they're waiting for.

Author: It naturally stops on its own. All of the commands stop on their own. 
How would you suggest we explicitly tell them what they're waiting for?

Reviewer: I suggest:

```suggestion
c) Start 4.x CouchDb and wait until it is up. You'll know it is up when the `docker-compose` call exits without errors:
```shell
cd ~/couchdb-single/ 
docker-compose up -d
cd ~/couchdb-migration/ 
docker-compose run couch-migration check-couchdb-up
```
```

Reviewer: I see this was added - resolving!

d) Change metadata to match the new CouchDb node
```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration move-node
```
e) Verify that the migration was successful
```shell
docker-compose run couch-migration verify

> ```

Reviewer: same comment as above - we should  qualify how they know `move-node` and `verify` calls correctly executed.

Author: Suggestions appreciated :) 
The command just completes without error at the moment. Should we have a success message at the end?

Reviewer: OMG - these two and `check-couchdb-up` all having a "$command complete" (or something) at the end would be dreamy!  If it's not too late, that'd be my vote.  Then we could have:

> Verify that the migration was successful burn running this and waiting for it to exit with `check-couchdb-up complete`

Otherwise, like above we did:

> Verify that the migration was successful burn running this `docker-compose` command and waiting for it to exit with no errors:`

Reviewer: I believe success messages were added in #medic/couchdb-migration/pull/7, so we can now update this section with the corresponding messages

Reviewer: i see these were added - thanks!

> If all checks pass, proceed with starting CHT-Core 4.x, using the same environment variables.   

Reviewer: ```suggestion
If all checks pass, proceed with starting CHT-Core 4.x, using the same environment variables you saved in `~/couchdb-single/.env`.   
```

#### Multi node

a) Download 4.x clustered CouchDb docker-compose file:
```shell
mkdir -p ~/couchdb-cluster/ 
cd ~/couchdb-cluster/ 
curl -s -o ./docker-compose.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.1.0/docker-compose/cht-couchdb-clustered.yml
```
b) Create a data folder for every one of the CouchDb nodes.

> 

Reviewer: Seems worth an example (though I had trouble formatting code inside a suggestion :disappointed: ) :

```
b) Create a data folder for every one of the CouchDb nodes. If you were going to a 3 cluster node, this would be:

   mkdir -p ~/couchdb-cluster/node1
   mkdir -p ~/couchdb-cluster/node2
   mkdir -p ~/couchdb-cluster/node3

```

(Ah - i see below you call it `couchdb_data_main` vs `node1`.  I like `main` as it's less ambiguous which is the main - which ever you want to standardize on!)

> c) Copy the 3.x CouchDb data folder into your main CouchDb node's folder. Your main node will be creating your cluster and adding the other secondary nodes to the cluster. You can tell which is your main node by checking your CouchDb docker compose files and identifying which service receives the `CLUSTER_PEER_IPS` environment variable. Your secondary nodes will receive `COUCHDB_SYNC_ADMINS_NODE` variable instead.  

Reviewer: This seems ambiguous to me:

> You can tell which is your main node by checking your CouchDb docker compose files

Given the importance of this step, I would switch to something like:

```suggestion
c) Copy the 3.x CouchDb data folder into `~/couchdb-cluster/node1`, which will be your main CouchDb node. This main node will create your cluster and the other secondary nodes will be added to it. In `node1`'s environment variable file, define  `CLUSTER_PEER_IPS` . In all other secondary nodes, declare the `COUCHDB_SYNC_ADMINS_NODE` variable instead.  
```

But I'm not sure if that's technically accurate!

Speaking of which - we should specify where you declare these variables.  It seems like each node has it's own `.env` file, but we've only made one in `~/couchdb-cluster/.env`  below. I don't know where to put `CLUSTER_PEER_IPS` vs where to put `COUCHDB_SYNC_ADMINS_NODE` .

Author: > It seems like each node has it's own .env file, 

This will depend entirely on how we instruct people to self host clusters, which is now unknown?

/couchdb_data_main
   /.delete
   /_dbs.couch
   /_nodes.couch
   /_users.couch
   /.shards
     /00000000-15555554
     /2aaaaaaa-3ffffffe
     /3fffffff-55555553
     /6aaaaaa9-7ffffffd
     /7ffffffe-95555552
     /15555555-2aaaaaa9
     /55555554-6aaaaaa8
     /95555553-aaaaaaa7
   /shards
     /00000000-15555554
     /2aaaaaaa-3ffffffe
     /3fffffff-55555553
     /6aaaaaa9-7ffffffd
     /7ffffffe-95555552
     /15555555-2aaaaaa9
     /55555554-6aaaaaa8   
     /95555553-aaaaaaa7
     
/couchdb_data_secondary
   /.shards   
   /shards
```
After moving two shards: `55555554-6aaaaaa8` and `6aaaaaa9-7ffffffd`
```shell
/couchdb_data_main
   /.delete
   /_dbs.couch
   /_nodes.couch
   /_users.couch
   /.shards
     /00000000-15555554
     /2aaaaaaa-3ffffffe
     /3fffffff-55555553     
     /7ffffffe-95555552
     /15555555-2aaaaaa9     
     /95555553-aaaaaaa7
   /shards
     /00000000-15555554
     /2aaaaaaa-3ffffffe
     /3fffffff-55555553     
     /7ffffffe-95555552
     /15555555-2aaaaaa9
     /95555553-aaaaaaa7
          
/couchdb_data_secondary
   /.shards   
     /6aaaaaa9-7ffffffd
     /55555554-6aaaaaa8
   /shards
     /6aaaaaa9-7ffffffd
     /55555554-6aaaaaa8
```

> i) Change metadata to match the new shard distribution. You will be using the same `shard_matrix` as the step above. 

Reviewer: ```suggestion
i) Change metadata to match the new shard distribution. We declared `$shard_matrix` in step "g" above, so it is still set now: 
```

---
title: "Migration from CHT 3.x to CHT 4.x"
linkTitle: "Data migration to 4.x"
weight: 10
description: >
  Guide to migrate existent data from CHT 3.x to CHT 4.x
aliases:
relatedContent: >
---

The hosting architecture differs entirely between CHT-Core 3.x and CHT-Core 4.x. Migrating data from an existing instance running CHT 3.x requires a few manual steps. 
This guide will present the required steps while using a migration helping tool, called `couchdb-migration`. This tool interfaces with CouchDb, to update shard maps and database metadata.
By the end of this guide, your CHT-Core 3.x CouchDb will be down and CHT-Core 4.x ready to be used.
Using this tool is not required, and the same result can be achieved by calling CouchDb endpoints directly. [Consult CouchDB documentation for details about moving shards](https://docs.couchdb.org/en/3.2.2-docs/cluster/sharding.html#moving-a-shard). 

### 1. Install CHT data migration tool

Open your terminal and run these commands. They will create a new directory, download a docker compose file and download the required docker image. 
```shell
mkdir -p ~/couchdb-migration/ 
cd ~/couchdb-migration/ 

> curl -s -o ./docker-compose.yml https://github.com/medic/couchdb-migration/blob/main/docker-compose.yml

Reviewer: We need to specify the raw URL:

```suggestion
curl -s -o ./docker-compose.yml https://raw.githubusercontent.com/medic/couchdb-migration/main/docker-compose.yml
```

docker-compose up
```

For the following steps, the tool needs access to your CouchDb installation. To allow this access, you will need to provide a URL to your CouchDB installation that includes authentication. 
Additionally, if your CouchDb runs in docker, the tool needs to be added to the same docker network in order to access protected endpoints:

```shell
export COUCH_URL=http://<authentication>@<host-ip>:<port>
```
or 
```shell
export CHT_NETWORK=<docker-network-name>
export COUCH_URL=http://<authentication>@<docker-service-name>:<port>
```

For simplicity, you could store these required values in an `.env` file: 
```shell
cat > ${HOME}/couchdb-migration/.env << EOF
CHT_NETWORK=<docker-network-name>
COUCH_URL=http://<authentication>@<docker-service-name>:<port>
EOF

> ```

Reviewer: Per our chat on Slack today, these all need to be `https`, correct?  I think it's worth updating the URLs to be `https` and explicitly call out that the CouchDB Migration utility needs to talk over `https` when talking to Medic OS.

[Consult information about backups for details]({{< relref "apps/guides/hosting/3.x/self-hosting#backup" >}}).
To ensure no changes happen to your CouchDB data in your CHT 3.x server, stop the API by getting shell on your Medic OS container and calling `/boot/svc-stop medic-api`.

To minimize downtime when upgrading, it's advised to prepare the 3.x installation for the 4.x upgrade, and pre-index all views that are required by 4.x.

The migration tool provides a command which will download all 4.x views to your 3.x CouchDb, and initiate view indexing. `<desired CHT version>` is any version at or above `4.0.0`:

```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration pre-index-views <desired CHT version>
```

Once view indexing is finished, proceed with the next step.

{{% alert title="Note" %}} If this step is omitted, 4.x API will fail to respond to requests until all views are indexed. Depending on the size of the database, this could take many hours, or even days. {{% /alert %}} 

### 3. Save existent CouchDb configuration

> Some CouchDb configuration values must be ported from existent CouchDb to the 4.x installation. Store them in a safe location before shutting down 3.x CouchDb.
> Use the migration tool to obtain these values:
> ```shell
> cd ~/couchdb-migration/ 
> docker-compose run couch-migration get-env
> ```
> 
> ##### a) CouchDB secret 
> Used in encrypting all CouchDb passwords and session tokens.
> ##### b) CouchDb server uuid
> Used in generating replication checkpointer documents, which track where replication progress between every client and the server, and ensure that clients don't re-download or re-upload documents.
> 

Reviewer: Having run through the guide now once (almost!), I think we can be a bit more explicit here:

```suggestion
Some CouchDb configuration values must be ported from existent CouchDb to the 4.x installation. Store them in a safe location before shutting down 3.x CouchDb.
Use the migration tool to obtain these values:
```shell
cd ~/couchdb-migration/ 
docker-compose run couch-migration get-env
```

Four values will be output. You already know the user (`COUCHDB_USER`) and password (`COUCHDB_PASSWORD`), but the two you need to save for later are:

* `COUCHDB_SECRET` -  CouchDB secret - Used in encrypting all CouchDB passwords and session tokens.
* `COUCHDB_UUID` -  `CouchDB server UUID - Used in generating replication checkpointer documents, which track where replication progress between every client and the server, and ensure that clients don't re-download or re-upload documents.
```

By the end of this guide, your CHT-Core 3.x CouchDb will be down and CHT-Core 4.x ready to be used.
Using this tool is not required, and the same result can be achieved by calling CouchDb endpoints directly. [Consult CouchDB documentation for details about moving shards](https://docs.couchdb.org/en/3.2.2-docs/cluster/sharding.html#moving-a-shard). 

### 1. Install CHT data migration tool

Open your terminal and run these commands. They will create a new directory, download a docker compose file and download the required docker image. 
```shell
mkdir -p ~/couchdb-migration/ 
cd ~/couchdb-migration/ 
curl -s -o ./docker-compose.yml https://raw.githubusercontent.com/medic/couchdb-migration/main/docker-compose.yml
docker-compose up
```

For the following steps, the tool needs access to your CouchDb installation. To allow this access, you will need to provide a URL to your CouchDB installation that includes authentication. 
If your installation exposes a different port for CouchDb cluster API endpoints, please export that port.
If running against an installation of `MedicOS`, please make sure that the protocol of the URL is `https`. 

> 

Reviewer 2: In b) you mean: don't define CHT_NETWORK and set COUCH_URL to https://<local_network_ip>
I think option b is more friendly.

EOF
```

c) Update `couchdb-migration` environment variables. Depending on your setup, it's possible you will need to update `CHT_NETWORK` and `COUCH_URL` to match the newly started 4.x CouchDb.
From this point on, the `couchdb-migration` container should connect to the same docker network as your CouchDb installation, in order to access APIs that are only available on protected ports. Correctly setting `CHT_NETWORK` is required for the next steps to succeed.
To get the correct `docker-network-name` and `docker-service-nam`, you can use `docker network ls` to list all networks and `docker network inspect <docker-network-name>` to get the name of the CouchDb container that exists in this network.

```shell
cat > ${HOME}/couchdb-migration/.env << EOF
CHT_NETWORK=<docker-network-name>
COUCH_URL=http://<authentication>@<docker-container-name>:<port>
EOF
```

d) Start 4.x CouchDb and wait until it is up. You'll know it is up when the `docker-compose` call exits without errors and logs `CouchDb is Ready`.  
```shell
cd ~/couchdb-single/ 
docker-compose up -d
cd ~/couchdb-migration/ 
docker-compose run couch-migration check-couchdb-up

> ```

Author: Changed.
